// rettiwt-api/dist/models/args/TweetArgs.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TweetMediaArgs = exports.TweetArgs = void 0;
// PACKAGES
var class_validator_1 = require("class-validator");
var rettiwt_core_1 = require("rettiwt-core");
/**
 * The arguments specifying the tweet to be posted.
 */
var TweetArgs = exports.TweetArgs = /** @class */ (function () {
    /**
     * @param tweet - The tweet arguments specifying the tweet.
     */
    function TweetArgs(tweet) {
        this.text = tweet.text;
        this.media = tweet.media ? tweet.media.map(function (item) { return new TweetMediaArgs(item); }) : undefined;
        // Validating this object
        var validationResult = (0, class_validator_1.validateSync)(this);
        // If validation error occured
        if (validationResult.length) {
            throw new rettiwt_core_1.DataValidationError(validationResult);
        }
    }
    __decorate([
        (0, class_validator_1.IsNotEmpty)(),
        (0, class_validator_1.IsString)(),
        (0, class_validator_1.MaxLength)(280),
        __metadata("design:type", String)
    ], TweetArgs.prototype, "text", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.IsArray)(),
        (0, class_validator_1.ArrayMaxSize)(4),
        (0, class_validator_1.IsObject)({ each: true }),
        __metadata("design:type", Array)
    ], TweetArgs.prototype, "media", void 0);
    return TweetArgs;
}());
/**
 * The arguments specifying the media to be posted in a single tweet.
 *
 * @public
 */
var TweetMediaArgs = exports.TweetMediaArgs = /** @class */ (function () {
    /**
     * @param media - The media arguments specifying the media.
     */
    function TweetMediaArgs(media) {
        var _a;
        this.path = media.path;
        this.tags = (_a = media.tags) !== null && _a !== void 0 ? _a : [];
        // Validating this object
        var validationResult = (0, class_validator_1.validateSync)(this);
        // If validation error occured
        if (validationResult.length) {
            throw new rettiwt_core_1.DataValidationError(validationResult);
        }
    }
    __decorate([
        (0, class_validator_1.IsNotEmpty)(),
        (0, class_validator_1.IsString)(),
        __metadata("design:type", String)
    ], TweetMediaArgs.prototype, "path", void 0);
    __decorate([
        (0, class_validator_1.IsOptional)(),
        (0, class_validator_1.IsArray)(),
        (0, class_validator_1.ArrayMaxSize)(10),
        (0, class_validator_1.IsNumberString)(undefined, { each: true }),
        __metadata("design:type", Array)
    ], TweetMediaArgs.prototype, "tags", void 0);
    return TweetMediaArgs;
}());
